src/org/javarosa/model/xform/SMSSerializingVisitor.java

```diff
@@ -133,83 +127,83 @@ public IDataPayload createSerializedPayload(FormInstance model, IDataReference r
     public void visit(FormInstance tree) {
         nodeSet = new String();
 
-        //TreeElement root = tree.getRoot();
-        TreeElement root = tree.resolveReference(rootRef);
+        TreeElement root = tree.getRoot();
 
         xmlns = root.getAttributeValue("", "xmlns");
         delimiter = root.getAttributeValue("", "delimiter");
-        if ( delimiter == null ) {
+        if (delimiter == null) {
             // for the spelling-impaired...
             delimiter = root.getAttributeValue("", "delimeter");
         }
         prefix = root.getAttributeValue("", "prefix");
 
-        xmlns = (xmlns != null)? xmlns : " ";
-        delimiter = (delimiter != null ) ? delimiter : " ";
+        xmlns = (xmlns != null) ? xmlns : " ";
+        delimiter = (delimiter != null) ? delimiter : " ";
         prefix = (prefix != null) ? prefix : " ";
 
-        //Don't bother adding any delimiters, yet. Delimiters are
-        //added before tags/data
-        theSmsStr = prefix;
+        theSmsStr = prefix.concat(delimiter);
 
-        // serialize each node to get it's answers
+        // serialize each node (and it's children) to get it's answers
+        serializeTree(root);
+        theSmsStr = theSmsStr.trim();
+    }
+
+    public void serializeTree(TreeElement root) {
         for (int j = 0; j < root.getNumChildren(); j++) {
-            TreeElement tee = root.getChildAt(j);
-            String e = serializeNode(tee);
-            if(e != null) {
-                theSmsStr += e;
+            TreeElement treeElement = root.getChildAt(j);
+            if (treeElement.isLeaf() && treeElement.getAttribute("", "tag") != null) {
+                String result = serializeNode(treeElement);
+                if (result != null) {
+                    theSmsStr += result;
```

Comment:
I think we should improve on repeatedly mutating this field here. Let the method be pure (not modify any fields). Here’s what I’m thinking: 
https://github.com/dcbriccetti/javarosa/commit/5f016622a1ed0523b9a5f54e301beb9bbd2f8101